### PR TITLE
Adding Vertical order Traversal of binary tree

### PR DESCRIPTION
**Pull request by Chahat-M**
## Related Issue
- #84 

## Proposed Changes
- Added vertical order traversal of a binary tree using level order traversal along with time complexity
- Updated README.md accordingly

## Additional Info
- Vertical order traversal is the method of traversing a tree from top to bottom and left to right.
- In the output, the node that is at leftmost horizontal distance of the root node is printed first.
- Then the other nodes that are at same horizontal distance are printed from top to bottom and left to right, ending with the rightmost node. 
- There are various methods for the same, but here it is being implemented using level order traversal.

## Checklist

- [x] ✅ My code follows the code style of this project.
- [x] 📝 My change requires a change to the documentation.
- [x] 🎀 I have updated the documentation accordingly.
- [x] 👀 I have read the [**CONTRIBUTING**](https://github.com/Lakhankumawat/LearnCPP/blob/main/Contributing.md) document.
- [x] ✨ I have added tests to cover my changes.
- [x] 🚩 All new and existing tests passed.
- [x] 🌟 ed the repo

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Input/Output Screenshots
image

### CODE REVIEW

**Review by Lakhankumawat — CHANGES_REQUESTED** [review]
Your Code is unefficient and hardcoded , please improve it !!.

**Comment by Chahat-M** [comment]
@Lakhankumawat Can you please clarify the above doubt, so that I can work accordingly on it again?

File: T-Tree/T-TreeTraversals/verticalTraversal.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    root->right = new Node(3);
+    root->left->left = new Node(4);
+    root->left->right = new Node(5);
+    root->right->left = new Node(6);
+    root->right->right = new Node(7);
+    root->right->left->right = new Node(8);
+    root->right->right->right = new Node(9);
+    root->right->right->left = new Node(10);
+    root->right->right->left->right = new Node(11);
+    root->right->right->left->right->right = new Node(12);
```
Comment (Lakhankumawat) [inline T-Tree/T-TreeTraversals/verticalTraversal.cpp:74]: this you are doing here ! , is not a good appraoch

Comment (Lakhankumawat) [reply T-Tree/T-TreeTraversals/verticalTraversal.cpp:74]: Instead you can use a for  or while loop to make it more efficient and elegant

**Review by Lakhankumawat — CHANGES_REQUESTED** [review]

**Comment by Lakhankumawat** [comment]
1. Edit Your PR descriptive and describe your PR in a bit of details .

**Comment by Lakhankumawat** [comment]
You need to follow particular guidelines for your README , please follow
these guidelines [click here](https://github.com/Lakhankumawat/LearnCPP#to-add-programs-using-stl-follow-same-approach-as--algorithms) 
also you can see how your README should look like [see here](https://github.com/Lakhankumawat/LearnCPP/blob/main/S-SortingAlgorithms/README.md) 
for more please go through these [here](https://youtu.be/7UgC3HDETRA)

**Comment by Lakhankumawat** [comment]
once you re done these changes and one of mentors reviews your PR , your PR will be merged , thanks .

File: T-Tree/T-TreeTraversals/verticalTraversal.cpp (hunk 1 above)
Comment (Chahat-M) [reply T-Tree/T-TreeTraversals/verticalTraversal.cpp:74]: I was thinking of asking the user to input inorder and preorder traversal of the binary tree to create the specific tree as desired? What are your views on this?

**Comment by Chahat-M** [comment]
Thanks for reviewing. I'll make the changes and update  soon.

**Comment by Chahat-M** [comment]
@Lakhankumawat I have updated the code and README accordingly. Are there any more changes to be done?

File: T-Tree/README.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+14 7 19 5 30 10 25 15
+Enter Preorder traversal of the tree: 
+5 7 14 19 10 30 15 25
+Vertical order traversal is 
+14
+7
+5 19 30
+10 25
+15
+```
```
Comment (siddhi-244) [inline T-Tree/README.md:34]: Add time complexity in readme

File: T-Tree/README.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+But before understanding Vertical order traversal, we should be aware with the term - *horizontal distance*.<br>
+**Horizontal distance** - Horizontal distance is the distance of the node from the root node. Thus,
+- Horizontal distance of root node = 0
+- Horizontal distance of left child = Horizontal distance of parent - 1
+- Horizontal distance of right child = Horizontal distance of parent + 1
+<br></br>
+<img src = "https://i.imgur.com/I7g4VHa.png"/>
+<br></br>
+#### In vertical order traversal, the nodes that are at the left most horizontal distance from the root are printed first and the nodes that are at the rightmost horizontal distance are printed at the end. <br> The nodes that lie at the same horizontal distance are then printed from top to bottom. <br> If the nodes are at same horizontal distance and same level, then they are printed from right to left. This is ensured in the code by using level order traversal approach. <br>
+The time complexity for the same is O(n * logn), where n is the no. of total nodes.
```
Comment (Chahat-M) [inline T-Tree/README.md:12]: I have already added the time complexity here.

Comment (siddhi-244) [reply T-Tree/README.md:12]: Try to include it in a separate heading. You can check other prs !

Comment (Chahat-M) [reply T-Tree/README.md]: Done!!

File: T-Tree/README.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+#### **Algorithm -** 
+1. A binary tree is constructed on the basis of the user input (number of nodes, inorder traversal, preorder traversal)<br>
+2. Horizontal distance of the nodes are calculated and stored alongwith the node in a queue.<br>
+3. All the nodes corresponding to the same horizontal distance are stored in a map.<br>
+4. To print the vertical order traversal, the map is traversed. 
+
+#### **Time Complexity -**
+The time complexity for the same is O(n * logn), where n is the no. of total nodes.
+
```
Comment (siddhi-244) [inline T-Tree/README.md:21]: Add space Complexity too

Comment (Chahat-M) [reply T-Tree/README.md:22]: Done!

**Review by siddhi-244 — CHANGES_REQUESTED** [review]
Follow the readme template.  Add table of contents

**Comment by siddhi-244** [comment]
> @siddhi-244 The table of contents should include the contents of Vertical order traversal or all the various traversals?

For the one which you have done !

**Comment by aditi944** [comment]
closed

**Review by siddhi-244 — APPROVED** [review]
LGTM

**Review by Lakhankumawat — APPROVED** [review]
## LGTM ✨💎

Your contribution is awesome @Chahat-M 🚩

Welcome to the team 🤝
Checkout your name on contributors list 🦖
Congo on your first pr being merged 🎀

**Comment by Lakhankumawat** [comment]
@all-contributors please add @Chahat-M code ,

**Pull request merged**